src/slice.rs

```diff
@@ -282,188 +303,323 @@ impl_sliceorindex_from_index!(isize);
+    Do: Dimension,
 {
     type Target = T;
     fn deref(&self) -> &Self::Target {
         &self.indices
     }
 }
 
-impl<T, D> SliceInfo<T, D>
+impl<T, Di, Do> SliceInfo<T, Di, Do>
 where
-    D: Dimension,
+    Di: Dimension,
+    Do: Dimension,
 {
     /// Returns a new `SliceInfo` instance.
     ///
-    /// If you call this method, you are guaranteeing that `out_dim` is
-    /// consistent with `indices`.
+    /// If you call this method, you are guaranteeing that `in_dim` and
+    /// `out_dim` are consistent with `indices`.
     #[doc(hidden)]
-    pub unsafe fn new_unchecked(indices: T, out_dim: PhantomData<D>) -> SliceInfo<T, D> {
+    pub unsafe fn new_unchecked(
+        indices: T,
+        in_dim: PhantomData<Di>,
+        out_dim: PhantomData<Do>,
+    ) -> SliceInfo<T, Di, Do> {
         SliceInfo {
+            in_dim: in_dim,
             out_dim: out_dim,
             indices: indices,
         }
     }
 }
 
-impl<T, D> SliceInfo<T, D>
+impl<T, Di, Do> SliceInfo<T, Di, Do>
 where
-    T: AsRef<[SliceOrIndex]>,
-    D: Dimension,
+    T: AsRef<[AxisSliceInfo]>,
+    Di: Dimension,
+    Do: Dimension,
 {
     /// Returns a new `SliceInfo` instance.
     ///
-    /// Errors if `D` is not consistent with `indices`.
-    pub fn new(indices: T) -> Result<SliceInfo<T, D>, ShapeError> {
-        if let Some(ndim) = D::NDIM {
-            if ndim != indices.as_ref().iter().filter(|s| s.is_slice()).count() {
+    /// Errors if `Di` or `Do` is not consistent with `indices`.
+    pub fn new(indices: T) -> Result<SliceInfo<T, Di, Do>, ShapeError> {
+        if let Some(ndim) = Di::NDIM {
+            if ndim != indices.as_ref().iter().filter(|s| !s.is_new_axis()).count() {
+                return Err(ShapeError::from_kind(ErrorKind::IncompatibleShape));
+            }
+        }
+        if let Some(ndim) = Do::NDIM {
+            if ndim != indices.as_ref().iter().filter(|s| !s.is_index()).count() {
                 return Err(ShapeError::from_kind(ErrorKind::IncompatibleShape));
             }
         }
         Ok(SliceInfo {
+            in_dim: PhantomData,
             out_dim: PhantomData,
             indices: indices,
         })
     }
 }
 
-impl<T: ?Sized, D> SliceInfo<T, D>
+impl<T: ?Sized, Di, Do> SliceInfo<T, Di, Do>
 where
-    T: AsRef<[SliceOrIndex]>,
-    D: Dimension,
+    T: AsRef<[AxisSliceInfo]>,
+    Di: Dimension,
+    Do: Dimension,
 {
+    /// Returns the number of dimensions of the input array for
+    /// [`.slice()`](struct.ArrayBase.html#method.slice).
+    ///
+    /// If `Di` is a fixed-size dimension type, then this is equivalent to
+    /// `Di::NDIM.unwrap()`. Otherwise, the value is calculated by iterating
+    /// over the `AxisSliceInfo` elements.
+    pub fn in_ndim(&self) -> usize {
+        Di::NDIM.unwrap_or_else(|| {
+            self.indices
+                .as_ref()
+                .iter()
+                .filter(|s| !s.is_new_axis())
+                .count()
+        })
```

Comment:
I think if let / match is much easier to read than unwrap_or_else.

Reply:
Yes, I agree. I have a tendency to use the methods on `Option` too much when pattern matching would work.